styx-api-service/src/main/java/com/spotify/styx/api/BackfillResource.java

```diff
@@ -152,20 +173,67 @@ private BackfillsPayload getBackfills(RequestContext requestContext) {
     }
   }
 
-  private Response<Void> haltBackfill(String id) {
+  private String schedulerApiUrl(CharSequence... parts) {
+    return schedulerServiceBaseUrl + SCHEDULER_BASE_PATH + "/" + String.join("/", parts);
+  }
+
+  private CompletionStage<Response<ByteString>> haltBackfill(String id, RequestContext rc) {
     try {
       final Optional<Backfill> backfillOptional = storage.backfill(id);
       if (backfillOptional.isPresent()) {
-        storage.storeBackfill(backfillOptional.get().builder().halted(true).build());
+        final Backfill backfill = backfillOptional.get();
+        storage.storeBackfill(backfill.builder().halted(true).build());
+        return haltActiveBackfillInstances(backfill, rc.requestScopedClient());
       } else {
-        return Response.forStatus(Status.NOT_FOUND.withReasonPhrase("backfill not found"));
+        return CompletableFuture.completedFuture(
+            Response.forStatus(Status.NOT_FOUND.withReasonPhrase("backfill not found")));
       }
     } catch (IOException e) {
-      return Response.forStatus(
+      return CompletableFuture.completedFuture(Response.forStatus(
           Status.INTERNAL_SERVER_ERROR
-              .withReasonPhrase("could not halt backfill: " + e.getMessage()));
+              .withReasonPhrase("could not halt backfill: " + e.getMessage())));
+    }
+  }
+
+  private CompletionStage<Response<ByteString>> haltActiveBackfillInstances(Backfill backfill, Client client) {
+    return CompletableFutures.allAsList(
+        retrieveBackfillStatuses(backfill).stream()
+            .filter(this::activeState)
+            .map(RunStateData::workflowInstance)
+            .map(workflowInstance -> haltActiveBackfillInstance(workflowInstance, client))
+            .collect(toList()))
+        .handle((result, throwable) -> {
+          if (throwable != null || result.contains(Boolean.FALSE)) {
+            return Response.forStatus(
+                Status.INTERNAL_SERVER_ERROR
+                    .withReasonPhrase(
+                        "some active instances cannot be halted"));
+          } else {
+            return Response.ok();
+          }
+        });
+  }
+
+  private CompletionStage<Boolean> haltActiveBackfillInstance(WorkflowInstance workflowInstance,
+                                                                Client client) {
+    try {
+      final ByteString payload = serialize(Event.halt(workflowInstance));
+      final Request request = Request.forUri(schedulerApiUrl("events"), "POST")
+          .withPayload(payload);
+      return client.send(request)
+          .thenApply(response -> response.status().family().equals(SUCCESSFUL));
+    } catch (JsonProcessingException e) {
+      return CompletableFuture.completedFuture(false);
+    }
+  }
+
+  private boolean activeState(RunStateData runStateData) {
```

Comment:
This function could be static, right? And maybe naming it `isActiveState` is more clear.